### PR TITLE
Eos 24915 2: Added Source Code for Encryption of Secret Key

### PR DESCRIPTION
**Pull request by kailashyadav11**
# Problem Statement
- Accept Secret from ConfigMap and Decrypt from within container
  Take Confidential value encrypt using cluster ID+Component ID and store onto conf store.

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [x] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [x] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [x] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

**Comment by cortx-admin** [comment]
Can one of the admins verify this patch?

File: src/provisioner/provisioner.py [hunk 1]
```diff
@@ -50,27 +50,36 @@ def config_apply(solution_conf_url: str, cortx_conf_url: str = None):
...
+                raise CortxProvisionerError(errno.EINVAL, 'Cluster ID not specified')
+            # TODO: Find a better way to store the encryption key
+            encryption_key = Cipher.gen_key(cluster_id, 'cluster')
+            with open('/etc/encrypt', 'wb') as en:
+                en.write(encryption_key)
+            
             CortxProvisioner.config_apply_cluster(cortx_config_store)
 
         if Conf.get(CortxProvisioner._solution_index, 'cortx') is not None:
+            # TODO: Loop through all the secret keys
```
Comment (nitin-seagate) [inline src/provisioner/provisioner.py:68]: Add logic to update all secrets as follows:
```
for key in Conf.get_keys(CortxProvisioner._solution_index):
            if key.endswith('secret'):
                secret_val = Conf.get(CortxProvisioner._solution_index, key)
                val = None
                with open(os.path.join('/etc/cortx/solution/secret', secret_val)) as f:
                    val = f.read()
                if val is None:
                    raise CortxProvisionerError(errno.EINVAL,
                        'Could not find the Secret in /etc/cortx/solution/secret')
                with open(CIPHER_KEY, 'rb') as cipher_obj:
                    cipher_key = cipher_obj.read()
                val = Cipher.encrypt(cipher_key, val.encode('ascii'))
                Conf.set(CortxProvisioner._solution_index, key, val)
```

File: src/provisioner/provisioner.py [hunk 2]
```diff
@@ -50,27 +50,36 @@ def config_apply(solution_conf_url: str, cortx_conf_url: str = None):
...
-                secret_val = Conf.get(CortxProvisioner._solution_index, key)
-                val = None
-                with open(os.path.join('/etc/cortx/solution/secret', secret_val)) as f:
-                    val = f.read()
-                if val is None:
-                    raise CortxProvisionerError(errno.EINVAL,
-                        'Could not find the Secret in /etc/cortx/solution/secret')
-                val = Cipher.encrypt(encryption_key, val.encode('ascii'))
-                Conf.set(CortxProvisioner._solution_index, key, val)
+        encryption_key = None
```
Comment (nitin-seagate) [inline src/provisioner/provisioner.py:53]: Change variable name:
`encryption_key` to `cipher_key`

**Review by nitin-seagate — CHANGES_REQUESTED** [review]

File: src/provisioner/provisioner.py [hunk 3]
```diff
@@ -50,27 +50,36 @@ def config_apply(solution_conf_url: str, cortx_conf_url: str = None):
...
 
         if Conf.get(CortxProvisioner._solution_index, 'cluster') is not None:
+            # generating encryption key
+            cluster_id = Conf.get(CortxProvisioner._solution_index, 'cluster>id')
+            if cluster_id is None:
+                raise CortxProvisionerError(errno.EINVAL, 'Cluster ID not specified')
```
Comment (ujjwalpl) [inline src/provisioner/provisioner.py:59]:
```
        cluster_id = Conf.get(CortxProvisioner._solution_index, 'cluster>id')
        if cluster_id is None:
            cluster_id = ConfigStore.get('cluster>id')
        if cluster_id is None:
            raise CortxProvisionerError(errno.EINVAL, 'Cluster ID not specified')
```

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: src/provisioner/provisioner.py [hunk 4]
```diff
@@ -50,27 +50,36 @@ def config_apply(solution_conf_url: str, cortx_conf_url: str = None):
...
+            # TODO: Find a better way to store the encryption key
+            encryption_key = Cipher.gen_key(cluster_id, 'cluster')
+            with open('/etc/encrypt', 'wb') as en:
```
[the remark this answers is not in the document]
Comment (ujjwalpl) [reply src/provisioner/provisioner.py:62]: @nitin-seagate who generates /etc/cipher ?

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: src/provisioner/provisioner.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (kailashyadav11) [reply src/provisioner/provisioner.py]: updated the code accordingly

File: src/provisioner/provisioner.py (hunk 1 above)
Comment (kailashyadav11) [reply src/provisioner/provisioner.py]: updated the code accordingly

File: src/provisioner/provisioner.py (hunk 3 above)
Comment (kailashyadav11) [reply src/provisioner/provisioner.py:63]: updated the code accordingly

File: src/provisioner/provisioner.py (hunk 2 above)
Comment (kailashyadav11) [reply src/provisioner/provisioner.py]: udpated teh codeaccordingly

**Review by ujjwalpl — APPROVED** [review]

File: src/provisioner/provisioner.py [hunk 5]
```diff
@@ -50,27 +53,37 @@ def config_apply(solution_conf_url: str, cortx_conf_url: str = None):
...
-                val = Cipher.encrypt(encryption_key, val.encode('ascii'))
-                Conf.set(CortxProvisioner._solution_index, key, val)
-
+        cipher_key = None
         if Conf.get(CortxProvisioner._solution_index, 'cluster') is not None:
+            # generating encryption key
+            cluster_id = Conf.get(CortxProvisioner._solution_index, 'cluster>id')
+            if cluster_id is None:
+                cluster_id = ConfigStore.get('cluster>id')
+            if cluster_id is None:
```
Comment (nitin-seagate) [inline src/provisioner/provisioner.py:62]: Move if block within if-block

File: test/deploy/k8s/cortx-secret.yml [hunk 6]
```diff
@@ -6,4 +6,7 @@ type: Opaque
 data:
   openldap-secret: MTIzNDUK
   kafka-secret: YWJjZAo=
-  consul-secret: cGEyMzQ1Cg==
\ No newline at end of file
+  consul-secret: cGEyMzQ1Cg==
+  common-secret: MTIzNDUK
+  auth-secret: YWJjZAo=
+  mgmt-secret: cGEyMzQ1Cg==
```
Comment (nitin-seagate) [inline test/deploy/k8s/cortx-secret.yml:12]: Add new line at the end of file

File: src/provisioner/provisioner.py [hunk 7]
```diff
@@ -50,27 +53,37 @@ def config_apply(solution_conf_url: str, cortx_conf_url: str = None):
...
+            # TODO: Find a better way to store the encryption key
+            cipher_key = Cipher.gen_key(cluster_id, 'cluster')
+            with open(CIPHER_KEY, 'wb') as cipher_obj:
+                cipher_obj.write(cipher_key)
             CortxProvisioner.config_apply_cluster(cortx_config_store)
 
         if Conf.get(CortxProvisioner._solution_index, 'cortx') is not None:
+            # TODO: use /etc/cortx/solution/secret to confirm secret 
+            for key in Conf.get_keys(CortxProvisioner._solution_index):
+            # TODO: use /etc/cortx/solution/secret to confirm secret
```
Comment (nitin-seagate) [inline src/provisioner/provisioner.py:73]: Initialize: 
`cipher_key = None`

File: src/provisioner/provisioner.py [hunk 8]
```diff
@@ -50,27 +53,37 @@ def config_apply(solution_conf_url: str, cortx_conf_url: str = None):
...
+                    secret_val = Conf.get(CortxProvisioner._solution_index, key)
+                    val = None
+                    with open(os.path.join(CortxProvisioner._secrets_path, secret_val)) as secret:
+                        val = secret.read()
+                    if val is None:
+                        raise CortxProvisionerError(errno.EINVAL,
+                            f'Could not find the Secret in  {CortxProvisioner._secrets_path}')
+                    with open(CIPHER_KEY, 'rb') as cipher_obj:
+                        cipher_key = cipher_obj.read()
+                    val = Cipher.encrypt(cipher_key, val.encode('ascii'))
```
Comment (nitin-seagate) [inline src/provisioner/provisioner.py:84]: Add if-block for cipher_key is not None:
The only encrypt and set into ConfStore

File: src/provisioner/provisioner.py (hunk 7 above)
Comment (kailashyadav11) [reply src/provisioner/provisioner.py:73]: It is initialized already
        cortx_config_store = ConfigStore(cortx_conf_url)
        # source code for encrypting and storing secret key
        cipher_key = None

File: src/provisioner/provisioner.py (hunk 5 above)
Comment (kailashyadav11) [reply src/provisioner/provisioner.py:62]: moved the block accordingly

File: src/provisioner/provisioner.py (hunk 8 above)
Comment (kailashyadav11) [reply src/provisioner/provisioner.py:84]: Added check for cipher_key

**Pull request merged**